### PR TITLE
HPCC-17644 Support for remote key filtering + counting/limits.

### PR DESCRIPTION
**Pull request by jakesmith**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:

Passes regression with forceDafilesrv on (NB: loadBlob not yet implemented, so there's a few that were specifically omitted with forceDafilesrv=false)

Blob implementation added - now all regression suite tests pass.

### CODE REVIEW

**Comment by jakesmith** [comment]
@ghalliday @richardkchapman - please re-review

NB: 1st PR is from yet unmerged issue HPCC-15675

Previous PR for reference: https://github.com/hpcc-systems/HPCC-Platform/pull/10016

I am rerunning regression suite with forceDafilesrv on.
NB: there are several tests which access index blobs which failed as unimplemented at the moment, I turn forceDafilesrv off for those cases.

**Comment by jakesmith** [comment]
Passes regression with forceDafilesrv on (NB: loadBlob not yet implemented, so there's a few that were specifically omitted with forceDafilesrv=false)

File: common/remote/sockfile.cpp [hunk 1]
```diff
@@ -4206,6 +4217,22 @@ class CRemoteFileServer : implements IRemoteFileServer, public CInterface
...
+    {
+        CriticalBlock block(sect);
+        if (handle<=0)
+            return false;
+        unsigned clientidx;
+        unsigned handleidx;
+        if (!findHandle(handle,clientidx,handleidx))
+            return false;
+        CRemoteClientHandler &client = clients.item(clientidx);
+        OpenFileInfo &_fileInfo = client.openFiles.element(handleidx); // NB: links members
```
Comment (ghalliday) [inline common/remote/sockfile.cpp]: style: We normally reserve _ prefix for parmaters to constructors (or functions) that are initialising member variables.

something like openFileInfo would be clearer.

File: common/remote/sockfile.cpp [hunk 2]
```diff
@@ -2683,14 +2683,11 @@ class CRemoteKeyManager : public CSimpleInterfaceOf<IKeyManager>
...
+                unsigned maxRecs = maxRecsPerRequest;
                 if (maxRecs && chooseNLimit)
                 {
-                    if (totalGot >= chooseNLimit)
-                        break;
-                    maxRecs = chooseNLimit - totalGot;
-                    if (maxRecs > maxRecsPerRequest)
-                        maxRecs = maxRecsPerRequest; // limit to
+                    if (totalGot + maxRecs > chooseNLimit)
+                        maxRecs = chooseNLimit - totalGot;
```
Comment (ghalliday) [inline common/remote/sockfile.cpp]: picky: Should have a cast to unsigned otherwise the windows compiler will complain.  i.e.

(unsigned)(chooseNLimit - totalGo);

**Review by ghalliday — COMMENTED** [review]
A couple of initial comments.  I'll rereview it now.

**Comment by jakesmith** [comment]
@ghalliday - NB: I have updated the base PR that this PR is a based on: gh-10023

File: thorlcr/activities/indexread/thindexreadslave.cpp [hunk 3]
```diff
@@ -241,10 +232,27 @@ class CIndexReadSlaveBase : public CSlaveActivity
...
+            IPartDescriptor &part = partDescs.item(p);
+            RemoteFilename rfn;
+            part.getFilename(0, rfn);
+            StringBuffer filePath;
+            rfn.getPath(filePath);
+
+            Owned<IDelayedFile> lfile = queryThor().queryFileCache().lookup(*this, logicalFilename, part);
+            Owned<IKeyManager> klManager;
+
+            bool remoteKey = !localKey && !seekGEOffset && (!rfn.isLocal() || getOptBool("forceDafilesrv"));
```
Comment (ghalliday) [inline thorlcr/activities/indexread/thindexreadslave.cpp]: Should this only be enabled if a flag is set to enable the new remote functionality?

**Comment by ghalliday** [comment]
@jakesmith a few comments

File: thorlcr/activities/indexread/thindexreadslave.cpp (hunk 3 above)
Comment (jakesmith) [inline thorlcr/activities/indexread/thindexreadslave.cpp]: hm, because I've been testing locally and it is never enabled unless I force I had been thinking it was never enabled without forceDafilesrv... silly.

Yes, there should be an option to specifically enable. I'll add.

File: common/remote/sockfile.cpp (hunk 2 above)
Comment (jakesmith) [inline common/remote/sockfile.cpp]: will add.

**Comment by jakesmith** [comment]
@ghalliday - push new commit based on your last comments.

I've rebased this now that HPCC-15675 has been merged.

If you're finished reviewing, I'll squash.

**Comment by jakesmith** [comment]
@ghalliday - I've added blob support (last) commit, please review that.

File: system/jhtree/jhtree.cpp [hunk 4]
```diff
@@ -659,8 +659,9 @@ class jhtree_decl CKeyLevelManager : implements IKeyManager, public CInterface
             if (!started)
             {
                 started = true;
-                segs.checkSize(keyedSize, keyName.get());
-                numsegs = segs.segMonitors.length();
+                numsegs = segs.ordinality();
+                if (numsegs)
+                    segs.checkSize(keyedSize, keyName.get());
```
Comment (jakesmith) [inline system/jhtree/jhtree.cpp]: @richardkchapman @ghalliday - I made this change (avoid running checkSize() in reset()) because I did not want to serialize over the segment monitors per BLOB request.
So in the BLOB case there are 0 segment monitors and checkSize() would fail (as it checks size against calculated size from seg. monitors).

I think the change is safe, as it only applies to a key manager who have not had segment monitors setup and for blob access only that is not necessary.

What do you think?

Comment (ghalliday) [inline system/jhtree/jhtree.cpp]: Seems reasonable to me.

**Comment by jakesmith** [comment]
@ghalliday - squashed - but please hold off from merging until I have completed full regression suite runs with the blob changes.

File: common/remote/sockfile.cpp [hunk 5]
```diff
@@ -2582,6 +2596,202 @@ void CRemoteFile::copyTo(IFile *dest, size32_t buffersize, ICopyFileProgress *pr
...
+        Owned<CRemoteFileIO> remoteIO = prepKeySend(sendBuffer, RFCreadfilteredindexblob, false);
+        sendBuffer.append(blobId);
+        MemoryBuffer replyBuffer;
+        remoteIO->sendRemoteCommand(sendBuffer, replyBuffer);
+        replyBuffer.read(blobSize);
+        const byte *blobData = replyBuffer.readDirect(blobSize);
+        activeBlobs.append(replyBuffer.detach()); // NB: don't need to retain size, but keep sz+data to avoid copy
+        return blobData;
+    }
+    virtual void releaseBlobs() override
```
Comment (ghalliday) [inline common/remote/sockfile.cpp]: This should be called from the destructor to ensure it doesn't leak on an exception.

File: common/remote/sockfile.cpp [hunk 6]
```diff
@@ -2582,6 +2596,202 @@ void CRemoteFile::copyTo(IFile *dest, size32_t buffersize, ICopyFileProgress *pr
...
+    bool first = true;
+    unsigned __int64 chooseNLimit = 0;
+    ConstPointerArray activeBlobs;
+
+    CRemoteFileIO *prepKeySend(MemoryBuffer &sendBuffer, RemoteFileCommandType cmd, bool segmentMonitors)
+    {
+        Owned<IFileIO> iFileIO = delayedFile->getFileIO();
+        if (!iFileIO)
+            throw MakeStringException(0, "CRemoteKeyManager: Failed to open key file: %s", filename.get());
+        Linked<CRemoteFileIO> remoteIO = QUERYINTERFACE(iFileIO.get(), CRemoteFileIO);
```
Comment (ghalliday) [inline common/remote/sockfile.cpp]: general: QUERYINTERFACE is not efficient.  We should have a general design discussion about the best pattern for avoiding dynamic casts whenever we can.  (not for this PR.)

File: common/remote/sockfile.cpp [hunk 7]
```diff
@@ -3970,37 +4255,64 @@ class CRemoteFileServer : implements IRemoteFileServer, public CInterface
         PROGLOG("Exited CRemoteFileServer");
 #endif
     }
+    bool lookupFileIOHandle(int handle, OpenFileInfo &fileInfo, unsigned newFlags=0)
+    {
+        CriticalBlock block(sect);
```
Comment (ghalliday) [inline common/remote/sockfile.cpp]: trivial: could move this down below the if()

File: common/remote/sockfile.cpp [hunk 8]
```diff
@@ -3970,37 +4255,64 @@ class CRemoteFileServer : implements IRemoteFileServer, public CInterface
...
+        client.previdx = handleidx;
+        return true;
+    }
 
     //MORE: The file handles should timeout after a while, and accessing an old (invalid handle)
     // should throw a different exception
-    bool checkFileIOHandle(MemoryBuffer &reply, int handle, IFileIO *&fileio, bool del=false)
+    bool checkFileIOHandle(int handle, IFileIO *&fileio, bool del=false)
     {
         CriticalBlock block(sect);
```
Comment (ghalliday) [inline common/remote/sockfile.cpp]: trivial: could move below the clear and if()

**Review by ghalliday — COMMENTED** [review]
Very close - a few final comments

File: common/remote/sockfile.cpp (hunk 5 above)
Comment (jakesmith) [inline common/remote/sockfile.cpp]: good spot, will fix.

File: common/remote/sockfile.cpp (hunk 7 above)
Comment (jakesmith) [inline common/remote/sockfile.cpp]: will change.

File: common/remote/sockfile.cpp (hunk 8 above)
Comment (jakesmith) [inline common/remote/sockfile.cpp]: will change.

File: common/remote/sockfile.cpp (hunk 5 above)
Comment (jakesmith) [inline common/remote/sockfile.cpp]: actually, the key manager is never persisted (only the file IO is), so a key manager comes into existence per blob request and therefore is destroyed each time, taking any blobs with it that might not have been explicitly released.

**Comment by jakesmith** [comment]
@ghalliday - regression suite passed.

**Comment by ghalliday** [comment]
@jakesmith choosen can be passed if the transform doesn't skip.  Do you want to make that change?

File: common/remote/sockfile.cpp (hunk 5 above)
Comment (ghalliday) [inline common/remote/sockfile.cpp]: This is the class that is in the client - isn't it persisted?

**Comment by jakesmith** [comment]
>@jakesmith choosen can be passed if the transform doesn't skip. Do you want to make that change?

I see that I can do that for indexread (by checking transformMayFilter()).
I am not clear on how/what should happen on indexcount and indexnormalize.

Index count has a getChooseNLimit() method, though I am not sure what the ECL looks like (?).
Is it safe to limit the matches coming back from the key to this limit unconditionally ?

Index normalize, it has a getChooseNLimit() and a transform, so presumably could skip, but it has no transformMayFilter() method.
Simiarly for index normalize, it has getChooseNLimit() and transforms (so could skip), but there is no transformMayFilter() .

I see that there is a 'TIRunfilteredtransform' flag, but that is not generated for in the case I tested with a transform that doesn't skip.

@ghalliday

File: common/remote/sockfile.cpp (hunk 5 above)
Comment (jakesmith) [inline common/remote/sockfile.cpp]: Sorry, yes this is the client and you're right.
Have added a call to releaseBlobs() in destructor now to ensure released if exception etc.

**Comment by jakesmith** [comment]
@ghalliday - squashed
